File: codewars/6-kyu/sum-of-digits-(digital-root)/solution.py
# https://www.codewars.com/kata/541c8630095125aba6000c00
# digital root: https://en.wikipedia.org/wiki/Digital_root

# solution 1: 497ms
def digital_root(n):
    n = str(n)
    while len(n) > 1:
        n = str(sum(map(int, n)))
    return int(n)

# solution 2: 479ms
def digital_root(n):
    if n < 10:
        return n 
    else:
        return digital_root(sum(map(int, str(n))))

File: codewars/6-kyu/sum-of-digits-(digital-root)/solution.rs
// https://www.codewars.com/kata/541c8630095125aba6000c00

// solution 1: 1225ms
fn digital_root(n: i64) -> i64 {
    let mut number: i64 = num;

    loop {
        // break the digits into vector of digits
        let digits: Vec<i64> = number 
            .to_string()
            .chars()
            .map(|d| d.to_digit(10).unwrap() as i64)
            .collect();
        
        // sum the digits
        number = digits.iter().sum();

        // match the length of digit
        match digits.len() {
            2.. => continue, 
            _   => return number,
        }
    }
}


// solution 2: 
fn digital_root(n: i64) -> i64 {
    if n < 10 {
        n 
    } else {
        digital_root(sum_digits(n))
    }
}

fn sum_digits(n: i64) -> i64 {
    let mut sum = 0;
    let mut n = n;

    while n != 0 {
        sum += n % 10;
        n = n / 10;
    }

    sum 
}


// solution 3: 
fn digital_root(n: i64) -> i64 {
    (n - 1) % 9 + 1
}